src/utils/sweepline.jl

```diff
@@ -0,0 +1,182 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+"""
+    bentleyottmann(segments)
+
+Compute pairwise intersections between n `segments`
+in O(n⋅log(n)) time using Bentley-Ottmann sweep line
+algorithm.
+
+Outputs a Dictionary of {Point, Vector{Tuple{Point, Point}}}
+where the key is each intersection point and the values are all
+pairs of segments that intersect at that point.
+
+## References
+
+* Bentley, J. L., & Ottmann, T. 1979. [Algorithms for reporting and counting geometric intersections]
+  (https://www.itseng.org/research/papers/topics/VLSI_Physical_Design_Automation/Physical_Verification/DRC/Geometric_Intersection_Problems/1979-Bentley.pdf)
+"""
+function bentleyottmann(segments)
+  # adjust vertices of segments
+  segs = map(segments) do s
+    a, b = extrema(s)
+    a > b ? reverse(s) : s
+  end
+
+  # retrieve relevant info
+  s = first(segs)
+  p = minimum(s)
+  P = typeof(p)
+  S = Tuple{P,P}
+
+  # initialization
+  𝒬 = BinaryTrees.AVLTree{P}()
+  𝒯 = BinaryTrees.AVLTree{S}()
+  ℒ = Dict{P,Vector{S}}()
+  𝒰 = Dict{P,Vector{S}}()
+  𝒞 = Dict{P,Vector{S}}()
+  lookup = Dict{S,Int}()
+  for (i, s) in enumerate(segs)
+    a, b = extrema(s)
+    BinaryTrees.insert!(𝒬, a)
+    BinaryTrees.insert!(𝒬, b)
+    haskey(ℒ, a) ? push!(ℒ[a], (a, b)) : (ℒ[a] = [(a, b)])
+    haskey(𝒰, b) ? push!(𝒰[b], (a, b)) : (𝒰[b] = [(a, b)])
+    haskey(ℒ, b) || (ℒ[b] = S[])
+    haskey(𝒰, a) || (𝒰[a] = S[])
+    haskey(𝒞, a) || (𝒞[a] = S[])
+    haskey(𝒞, b) || (𝒞[b] = S[])
+    lookup[(a, b)] = i
+  end
+
+  # sweep line
+  points = Vector{P}()
+  segs = Vector{Vector{Int}}()
+  while !isnothing(BinaryTrees.root(𝒬))
+    p = BinaryTrees.key(BinaryTrees.minnode(𝒬))
+    BinaryTrees.delete!(𝒬, p)
+    handle!(points, segs, lookup, p, S, 𝒬, 𝒯, ℒ, 𝒰, 𝒞)
+  end
+  points, segs
+end
+
+function handle!(points, segs, lookup, p, S, 𝒬, 𝒯, ℒ, 𝒰, 𝒞)
+  ℬ = get(ℒ, p, S[])
+  ℰ = get(𝒰, p, S[])
+  ℐ = get(𝒞, p, S[])
+  _processend!(ℰ, 𝒬, 𝒯, 𝒞)
+  _processbegin!(ℬ, 𝒬, 𝒯, 𝒞)
+  _processintersects!(ℐ, 𝒬, 𝒯, 𝒞)
+  if !isempty(ℬ ∪ ℰ ∪ ℐ)
+    segments = ℬ ∪ ℰ ∪ ℐ
+    push!(points, p)
+    push!(segs, _pushintersection(lookup, segments))
+  end
+end
+
+function _processbegin!(ℬ, 𝒬, 𝒯, 𝒞)
+  [BinaryTrees.insert!(𝒯, s) for s in ℬ]
```

Comment:
```suggestion
  for s in ℬ
    BinaryTrees.insert!(𝒯, s)
  end
```
Array comprehension allocates a new object. Avoid this by using a for loop.